### PR TITLE
[memory][context][api][refactor] Unify names of ScMemoryContext methods

### PR DESCRIPTION
**Pull request by NikitaZotov**
* [x] Read PR [documentation](https://github.com/ostis-ai/sc-machine/blob/main/CONTRIBUTING.md)
* [x] Update changelog
* [x] Update documentation

Renaming C methods will be in a separate PR.

### CODE REVIEW

File: docs/changelog.md [hunk 1]
```diff
@@ -16,17 +16,43 @@ encapsulated this logic;
...
+  | FindLinksContentsByContentSubstring | SearchLinksContentsByContentSubstring |
+  | HelperSetSystemIdtf                 | SetElementSystemIdentifier            |
+  | HelperGetSystemIdtf                 | GetElementSystemIdentifier            |
+  | HelperResolveSystemIdtf             | ResolveElementSystemIdentifier        |
+  | HelperFindBySystemIdtf              | SearchElementBySystemIdentifier       |
+  | HelperGenTemplate                   | GenerateByTemplate                    |
+  | HelperSearchTemplate                | SearchByTemplate                      |
+  | HelperBuildTemplate                 | BuildByTemplate                       |
+  | CalculateStat                       | CalculateStatistics                   |
+  | BeingEventsPending                  | BeginEventsPending                    |
```
Comment (kilativ-dotcom) [inline docs/changelog.md:49]: What about event blocking?

Comment (NikitaZotov) [reply docs/changelog.md:49]: It was added in the current version

**Review by kilativ-dotcom — CHANGES_REQUESTED** [review]

File: docs/sc-memory/api/cpp/extended/agents/agent_context.md [hunk 2]
```diff
@@ -25,15 +25,15 @@ This method can be useful when you want generate sc-event subscription, but don'
 
 ```cpp
 ...
-// Create or use existing sc-agent context.
+// Generate or use existing sc-agent context.
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply docs/sc-memory/api/cpp/extended/agents/agent_context.md:28]: Okay

File: sc-memory/sc-memory/sc_memory.hpp [hunk 3]
```diff
@@ -153,86 +158,161 @@ class ScMemoryContext
...
+   * @throws ExceptionInvalidState if the sc-memory context is not authenticated or does not have read permissions.
+   *
+   * @code
+   * ScMemoryContext context;
+   * ScAddr elementAddr = context.GenerateNode(ScType::NodeConst);
+   * size_t outgoingArcsCount = context.GetElementOutgoingArcsCount(elementAddr);
+   * std::cout << "Outgoing arcs count: " << outgoingArcsCount << std::endl;
+   * @endcode
+   */
+  _SC_EXTERN size_t GetElementOutgoingArcsCount(ScAddr const & elementAddr) const noexcept(false);
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-memory/sc_memory.hpp:196]: Yes, I think we should fix it. These methods should only return the number of outgoing and incoming arcs, and we should have a method to get the number of edges

File: sc-memory/sc-memory/sc_memory.hpp [hunk 4]
```diff
@@ -1058,32 +1804,83 @@ class ScMemoryContext
...
+      ScTemplateSearchResultCallback const & callback,
+      ScTemplateSearchResultFilterCallback const & filterCallback = {},
+      ScTemplateSearchResultCheckCallback const & checkCallback = {}) noexcept(false);
+
+  /*!
+   * Searches sc-constructions by object of `ScTemplate` and passes found sc-constructions to `callback`
+   * lambda-function. If `filterCallback` passed, then all found constructions triples are filtered by `filterCallback`
+   * condition.
+   * @param templateToFind An object of `ScTemplate` to find sc-constructions by it.
+   * @param callback A lambda-function, callable when each sc-construction triple was found.
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-memory/sc_memory.hpp:1842]: No, it is called when any triple was found

File: sc-memory/sc-memory/sc_memory.hpp [hunk 5]
```diff
@@ -1058,32 +1804,83 @@ class ScMemoryContext
...
+   * @param filterCallback A lambda-function, that filters all found sc-constructions triples.
+   * @param checkCallback A lambda-function, that filters all found triples by checking single sc-elements.
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-memory/sc_memory.hpp:1844]: No, Filtering occurs by finding any matching triple by its elements

File: sc-memory/tests/sc-memory/common/test_sc_elements.cpp [hunk 6]
```diff
@@ -144,169 +142,169 @@ TEST_F(ScMemoryTest, CreateDeleteCountArcs)
...
+  EXPECT_EQ(ctx.GetElementOutgoingArcsCount(link), 0u);
+  EXPECT_EQ(ctx.GetElementIncomingArcsCount(link), 2u);
 }
 
 TEST_F(ScMemoryTest, CreateDeleteCountArcs2)
 {
   ScMemoryContext ctx;
 
-  ScAddr const node = ctx.CreateNode(ScType::Const);
+  ScAddr const node = ctx.GenerateNode(ScType::Const);
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/tests/sc-memory/common/test_sc_elements.cpp:189]: I'm not adding any new functionality in this pull request, the GenerateElement method can be done in a separate pull request

File: sc-memory/sc-memory/sc_memory.hpp [hunk 7]
```diff
@@ -688,68 +1063,55 @@ class ScMemoryContext
...
    * @param param2 The second parameter for the iterator.
    * @param param3 The third parameter for the iterator.
    * @param param4 The fourth parameter for the iterator.
    * @param param5 The fifth parameter for the iterator.
    * @param fn The function to be called for each result.
    *
-   * @note fn function should have 4 parameters (ScAddr const & source, ScAddr const & edge, ScAddr const & target,
-   * ScAddr const & attrEdge, ScAddr const & attr)
+   * @note fn function should have 5 parameters (ScAddr/ScType const & source, ScAddr/ScType const & connector,
+   * ScAddr/ScType const & target, ScAddr/ScType const & attrConnector, ScAddr/ScType const & attr)
```
[the remark this answers is not in the document]
Comment (NikitaZotov) [reply sc-memory/sc-memory/sc_memory.hpp:1114]: No, params can be types

[the remark this answers is not in the document]
Comment (kilativ-dotcom) [reply sc-memory/sc-memory/sc_memory.hpp:1114]: params for iterator can be types, but these lambdas are called for each found iterator so it takes addresses

File: sc-memory/sc-memory/sc_memory.hpp (hunk 4 above)
[the remark this answers is not in the document]
Comment (kilativ-dotcom) [reply sc-memory/sc-memory/sc_memory.hpp:1842]: sc-template-search.cpp:923 has comment
// there are no next triples for current triple, it is last
and after that check m_filterCallback is called, so I think that it is called to filter whole template, not single triple
same with m_callback and m_callbackWithRequest

File: docs/changelog.md (hunk 1 above)
Comment (kilativ-dotcom) [reply docs/changelog.md:55]: did you forget to commit line with event blocking?

File: docs/sc-memory/api/cpp/extended/template_api.md [hunk 8]
```diff
@@ -903,15 +903,15 @@ templ.Triple(
...
+bool const isFoundByTemplate = context.SearchByTemplate(templ, result);
 
 result.ForEach([](ScTemplateResultItem const & item) {
   // Implement logic to handle found sc-constructions.
 });
 ...
 ```
 
-## **HelperSmartSearchTemplate**
+## **SearchByTemplateWithControl**
```
[the remark this answers is not in the document]
Comment (kilativ-dotcom) [reply docs/sc-memory/api/cpp/extended/template_api.md:914]: search by template interruptible ?

File: docs/changelog.md [hunk 9]
```diff
@@ -16,17 +16,49 @@ encapsulated this logic;
...
   - improved the API and aligned with our description of how it should be;
   - simplified the API. Now it will be much easier to create agents, go right now and see how to work with the new API -- [C++ Agents API](sc-memory/api/cpp/extended/agents/agents.md).
+- All `ScMemoryContext` methods were redesigned to a common style. All methods with not correct names were deprecated, and new ones were added. See the table below to see which methods have been replaced.
+  
+  | Deprecated method                   | Substitution method                   |
+  |-------------------------------------|---------------------------------------|
+  | CreateNode                          | GenerateNode                          |
+  | CreateLink                          | GenerateLink                          |
+  | CreateEdge                          | GenerateConnector                     |
+  | GetElementOutputArcsCount           | GetElementOutgoingArcsCount           |
```
Comment (KovalM) [inline docs/changelog.md:34]:
```suggestion
  | GetElementOutputArcsCount           | GetElementOutgoingConnectorsCount           |
```
?

File: docs/changelog.md [hunk 10]
```diff
@@ -16,17 +16,49 @@ encapsulated this logic;
...
+  | GetElementInputArcsCount            | GetElementIncomingArcsCount           |
+  | GetEdgeSource                       | GetArcSourceElement                   |
+  | GetEdgeTarget                       | GetArcTargetElement                   |
+  | GetEdgeInfo                         | GetConnectorIncidentElements          |
+  | HelperCheckEdge                     | CheckConnector                        |
+  | FindLinksByContent                  | SearchLinksByContent                  |
+  | FindLinksByContentSubstring         | SearchLinksByContentSubstring         |
+  | FindLinksContentsByContentSubstring | SearchLinksContentsByContentSubstring |
+  | HelperSetSystemIdtf                 | SetElementSystemIdentifier            |
+  | HelperGetSystemIdtf                 | GetElementSystemIdentifier            |
```
Comment (KovalM) [inline docs/changelog.md:44]:
```suggestion
  | HelperGetSystemIdtf                 | GetElementSystemIdtf            |
```
Idtf is popular word

File: sc-memory/sc-memory/sc_iterator.cpp [hunk 11]
```diff
@@ -18,7 +18,7 @@ sc_iterator3 * CreateIterator3<sc_addr, sc_type, sc_addr>(
 }
 
 template <>
-sc_iterator3 * CreateIterator3<sc_addr, sc_type, sc_type>(
+sc_iterator3 * GenerateIterator3<sc_addr, sc_type, sc_type>(
```
Comment (KovalM) [inline sc-memory/sc-memory/sc_iterator.cpp:21]: 'Generate' has meaning 'generate in knowledge base'. Iterator doesn't have representation in KB

**Review by KovalM — CHANGES_REQUESTED** [review]

File: sc-memory/sc-memory/sc_memory.hpp (hunk 5 above)
[the remark this answers is not in the document]
Comment (kilativ-dotcom) [reply sc-memory/sc-memory/sc_memory.hpp:1844]: A lambda-function, that filters all found triples by checking each triple sc-element.

**Review by kilativ-dotcom — COMMENTED** [review]
not all comments were answered

**Review by KovalM — APPROVED** [review]

File: sc-memory/sc-core/sc-store/sc_iterator3.h [hunk 12]
```diff
@@ -125,7 +125,7 @@ _SC_EXTERN sc_iterator3 * sc_iterator3_f_f_f_new(
     sc_addr edge_addr,
     sc_addr end_addr);
 
-/*! Create new sc-iterator-3
+/*! Generate new sc-iterator-3
```
Comment (kilativ-dotcom) [inline sc-memory/sc-core/sc-store/sc_iterator3.h:128]: bring create back

File: sc-memory/sc-memory/sc_template_build.cpp [hunk 13]
```diff
@@ -118,47 +118,47 @@ class ScTemplateBuilder
...
       ScAddr const & objAddr = iter->Get(2);
 
       auto const & it = m_elements.find(objAddr.Hash());
       ObjectInfo obj = it == m_elements.cend() ? CollectObjectInfo(objAddr) : it->second;
 
-      ScAddr objSrc, objTrg;
-      if (obj.IsEdge() && m_context.GetEdgeInfo(objAddr, objSrc, objTrg))
+      if (obj.IsEdge())
       {
+        auto [objSrc, objTrg] = m_context.GetConnectorIncidentElements(objAddr);
```
Comment (kilativ-dotcom) [inline sc-memory/sc-memory/sc_template_build.cpp:134]: don't you agree that not all ScAddr variables should end in addr?

**Review by kilativ-dotcom — APPROVED** [review]

**Pull request merged**